content/blog/prototyping/index.md

```diff
@@ -0,0 +1,726 @@
+
+I will just scaffold out the functions or a module and then fill in the blanks later.
+It's like sketching: you don't have to fill in all the details right away and you can come back to it later once the composition is clear.
+
+```rust
+// We don't know yet how to process the data
+// but we're pretty certain that we need a function
+// that takes a Vec<i32> and returns an i32
+fn process_data(data: Vec<i32>) -> i32 {
+    todo!()
+}
+
+// There exists a function that loads the data and returns a Vec<i32>
+// How exactly it does that is not important right now
+fn load_data() -> Vec<i32> {
+    todo!()
+}
+
+fn main() {
+    // Given that we have a function to load the data
+    let data = load_data();
+    // ... and a function to process it
+    let result = process_data(data);
+    // ... we can print the result
+    println!("Result: {}", result);
+}
+```
+
+We did not do much here, but we have a clear idea of what the program should do.
+Now we can go and iterate on the design.
+For example, should `process_data` take a reference to the data?
+Should we create a struct to hold the data and the processing logic?
+How about using an iterator instead of a vector?
+Should we introduce a trait to support algorithms for processing the data?
+
+These are all helpful questions that we can answer without having to worry about the details of the implementation. 
+And yet our code is typesafe and compiles, and it is ready for refactoring. 
+
+### Avoid Generics
+
+Chances are, you might not know which parts of your application should be generic in the beginning.
+Therefore it's better to be conservative and use concrete types instead of generics until necessary. 
+
+So instead of writing this:
+
+```rust
+fn foo<T>(x: T) -> T {
+    x
+}
+```
+
+Write this:
+
+```rust
+fn foo(x: i32) -> i32 {
+    x
+}
+```
+
+If you need the same function for a different type, feel free to just copy and paste the function and change the type. 
+This way, you avoid the trap of settling on the wrong kind of abstraction too early.
+Maybe the two functions only differ by type signature for now, but they might serve a completely different purpose.
+If the function is not generic from the start, it's easier to remove the duplication later. 
+
+Only introduce generics when you see a clear pattern emerge in multiple places.
+
+Also avoid "fancy" generic type signatures:
+
+```rust
+fn foo<T: AsRef<str>>(x: T) -> String {
+    x.as_ref().to_string()
+}
+```
+
+Yes, this allows you to pass in a `&str` or a `String`, but it's also harder to read.
+
+Better use an owned type for your first implementation:
+
+```rust
+fn foo(x: String) -> String {
+    x
+}
+```
+
+Chances are, you won't need the flexibility at all.
+
+In summary, generics are powerful, but they can make the code harder to read and write.
+Stay clear of them until you have a good idea of what you're doing. 
```

Comment:
```suggestion
Avoid them until you have a clear idea of what you're doing. 
```

I often think my ideas are good. They are mine after all! But it much more objective to classify them as clear or not.